codegeex.el

```diff
@@ -163,8 +244,126 @@ The result will be replaced into the selected region."
+
+(defvar codegeex--post-command-timer nil)
+
+(defcustom codegeex-disable-predicates nil
+  "A list of predicate functions with no argument to disable Codegeex.
+Codegeex will not be triggered if any predicate returns t."
+  :type '(repeat function)
+  :group 'codegeex)
+
+(defcustom codegeex-enable-predicates '(evil-insert-state-p codegeex--buffer-changed)
+  "A list of predicate functions with no argument to enable Codegeex.
+Codegeex will be triggered only if all predicates return t."
+  :type '(repeat function)
+  :group 'codegeex)
+
+(defcustom codegeex-disable-display-predicates nil
+  "A list of predicate functions with no argument to disable Codegeex.
+Codegeex will not show completions if any predicate returns t."
+  :type '(repeat function)
+  :group 'codegeex)
+
+(defcustom codegeex-enable-display-predicates nil
+  "A list of predicate functions with no argument to enable Codegeex.
+Codegeex will show completions only if all predicates return t."
+  :type '(repeat function)
+  :group 'codegeex)
+
+(defmacro codegeex--satisfy-predicates (enable disable)
+  "Return t if satisfy all predicates in ENABLE and none in DISABLE."
+  `(and (cl-every (lambda (pred)
+                    (if (functionp pred) (funcall pred) t))
+                  ,enable)
+        (cl-notany (lambda (pred)
+                     (if (functionp pred) (funcall pred) nil))
+                   ,disable)))
+
+(defun codegeex--satisfy-trigger-predicates ()
+  "Return t if all trigger predicates are satisfied."
+  (codegeex--satisfy-predicates codegeex-enable-predicates codegeex-disable-predicates))
+
+(defun codegeex--satisfy-display-predicates ()
+  "Return t if all display predicates are satisfied."
+  (codegeex--satisfy-predicates codegeex-enable-display-predicates codegeex-disable-display-predicates))
+
+(defvar codegeex-mode-map (make-sparse-keymap)
+  "Keymap for Codegeex minor mode.
+Use this for custom bindings in `codegeex-mode'.")
+
+(defun codegeex--mode-enter ()
+  "Set up codegeex mode when entering."
+  (add-hook 'post-command-hook #'codegeex--post-command nil 'local))
+
+(defun codegeex--mode-exit ()
+  "Clean up codegeex mode when exiting."
+  (remove-hook 'post-command-hook #'codegeex--post-command 'local))
+
+;;;###autoload
+(define-minor-mode codegeex-mode
+  "Minor mode for Codegeex."
+  :init-value nil
+  :lighter " Codegeex"
+  (codegeex-clear-overlay)
+  (advice-add 'posn-at-point :before-until #'codegeex--posn-advice)
+  (if codegeex-mode
+      (codegeex--mode-enter)
+    (codegeex--mode-exit)))
+
+(defun codegeex--posn-advice (&rest args)
+  "Remap posn if in codegeex-mode."
+  (when codegeex-mode
+    (let ((pos (or (car-safe args) (point))))
+      (when (and codegeex--real-posn
+                 (eq pos (car codegeex--real-posn)))
+        (cdr codegeex--real-posn)))))
+
 ;;;###autoload
-(define-key global-map (kbd "M-\\") 'codegeex-buffer-completion)
+(define-global-minor-mode global-codegeex-mode
+  codegeex-mode codegeex-mode)
```

Comment:
user can use `codegeex-complete-at-point for "one shot" completion. Or use the `codegeex-mode` to have completion proposition inline when pausing the cursor